lib/nats.js

```diff
@@ -713,9 +713,13 @@ Client.prototype.setupHandlers = function () {
     return
   }
 
-  if (this.options.timeout) {
+  // tls upgrade will re-bind the handlers here, so if we have a timer we are going to ignore it
+  if (this.options.timeout && !this.connectionTimeoutHandler) {
     this.connectionTimeoutHandler = setTimeout(() => {
-      stream.destroy(new NatsError(CONN_TIMEOUT_MSG, CONN_TIMEOUT))
+      if (this.stream) {
+        // this fires on the current stream
+        this.stream.destroy(new NatsError(CONN_TIMEOUT_MSG, CONN_TIMEOUT))
+      }
```

Comment:
As we discussed, should we also delete timer on close stream?

Reply:
yes I refactored it to its own function